allow "IN" Expression to have a list or a set as a parameter

omiz opened the pull request:
this is because for values like Int64 passing an array to an IN expression is not allowed and requires variadic parameter

omiz committed:
allow "IN" Expression to have a list or a set as a parameter

this is because for values like Int64 passing an array to an IN expression is not allowed and requires variadic parameter

ftchirou reviewed (commented):
Hi @omiz! Thank you for your contribution. The changes look good to me. Before I merge them, could you cover them with some unit tests following the pattern [here](https://github.com/ftchirou/PredicateKit/blob/main/PredicateKitTests/OperatorTests.swift#L1317)? And maybe add a note in the [README](https://github.com/ftchirou/PredicateKit#in)?

omiz committed:
added tests for 'IN' expression with array and set arguments

omiz committed:
update README file with examples for "IN" with an array or a set

omiz commented:
I added the tests and updated the README file as requested to cover the new "IN" functions.
If there is anything else still missing before the merge please let me know so I can fix it

ftchirou commented on PredicateKitTests/OperatorTests.swift:
```diff
@@ -1335,6 +1335,49 @@ final class OperatorTests: XCTestCase {
     XCTAssertEqual(value, ["hello", "world", "welcome"])
   }
+    
+  func testKeyPathInArray() throws {
+    let predicate: Predicate<Data> = (\Data.text).in(["hello", "world", "welcome"])
+
+    guard case let .comparison(comparison) = predicate else {
+      XCTFail("text.in(['hello', 'world', 'welcome']) should result in a comparison")
+      return
+    }
+
+    guard let keyPath = comparison.expression.as(KeyPath<Data, String>.self) else {
+      XCTFail("the left side of the comparison should be a key path expression")
+      return
+    }
+
+    let value = try XCTUnwrap(comparison.value as? [String])
+
+    XCTAssertEqual(keyPath, \Data.text)
+    XCTAssertEqual(comparison.operator, .in)
+    XCTAssertEqual(value, ["hello", "world", "welcome"])
+  }
```
I think this is testing [this pre-existing function](https://github.com/ftchirou/PredicateKit/blob/main/PredicateKit/Predicate.swift#L483). That would explain why the [code coverage](https://github.com/ftchirou/PredicateKit/suites/5737924361/artifacts/190255986) went down. I would suggest to test on a different property. Something like:

```swift
let predicate: Predicate<Data> = (\Data.count).in([21, 42, 63])
/// ...
```

ftchirou commented on README.md:
````diff
@@ -348,6 +348,19 @@ You can use the  `in`  function to determine whether a property's value is one o
 // Matches all notes where the text is one of the elements in the specified list.
 let predicate = (\Note.text).in("a", "b", "c", "d")
 ```
+or pass in an Array
+
+```swift
+// Matches all notes where the text is one of the elements in the specified list.
+let predicate = (\Note.text).in(["a", "b", "c", "d"])
+```
+
+or pass in a Set 
+
+```swift
+// Matches all notes where the text is one of the elements in the specified list.
+let predicate = (\Note.text).in(Set(["a", "b", "c", "d"]))
+```
````
Small tweak to your changes. Let's use the following for the `in` section of the documentation.

###### in

You can use the  `in`  function to determine whether a property's value is one of the values in a comma-separated list, an array, or a set.


```swift
// Matches all notes where the text is one of the elements in the specified comma-separated list.
 let predicate = (\Note.text).in("a", "b", "c", "d")

// Matches all notes where the text is one of the elements in the specified array.
let predicate = (\Note.text).in(["a", "b", "c", "d"])

// Matches all notes where the text is one of the elements in the specified set.
let predicate = (\Note.text).in(Set(["a", "b", "c", "d"]))
```

When the property is of type `String`, `in` accepts a second parameter that determines how the string should be compared to the elements in the list.

```swift
// Case-insensitive comparison.
let predicate = (\Note.text).in(["a", "b", "c", "d"], .caseInsensitive)
```

ftchirou reviewed (commented):
Thank you for the updates. The [code coverage](https://github.com/ftchirou/PredicateKit/suites/5737924361/artifacts/190255986) went down. I added a suggestion to fix that and another suggestion for the README.

omiz committed:
update the README file for better description of 'IN' function

omiz committed:
fix 'in' tests to use to correct function for the array and set

omiz committed:
Fix ambiguous description of the in function with second parameter

omiz commented:
Thank you @ftchirou  for your suggestions. I tried to follow them as close as possible to what you mentioned

ftchirou reviewed (approved):
Looking good. 🎉
